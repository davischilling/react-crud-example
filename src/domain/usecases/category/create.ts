import { api } from '../../api';
import { ValidationError } from '../../errors/validation-error';
import { Category } from '../../models/category';
import { StatefulUseCase } from '../stateful-usecase';

export type State = {
  category: Partial<Category>;
};

export const DEFAULT_STATE: State = {
  category: {
    name: '',
    description: '',
    is_active: true,
  } as Partial<Category>,
};

export class CreateCategoryUseCase extends StatefulUseCase<State> {
  init = async () => {
    // no need for implementation
  };

  setCategory = ({ name, value }: { name: 'name' | 'description'; value: string }) => {
    if (!['name', 'description'].includes(name)) {
      throw new ValidationError(`Invalid attribute type: ${name}`);
    }
    this.setState({
      category: {
        ...this.state.category,
        [name]: value,
      },
    });
  };

  toggleIsActive = () => {
    this.setState({
      category: {
        ...this.state.category,
        is_active: !this.state.category.is_active,
      },
    });
  };

  createCategory = async () => {
    const fmtCategory: Partial<Category> = {
      ...this.getState().category,
      created_at: new Date().toString(),
      updated_at: new Date().toString(),
      deleted_at: null,
    };
    await api.post('/categories', fmtCategory);
  };
}
