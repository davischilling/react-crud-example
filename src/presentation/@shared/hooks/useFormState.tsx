import { useSnackbar } from 'notistack';
import { ChangeEvent, FormEvent, useMemo } from 'react';

type FormStateProps<Name> = {
  setEntity: ({ name, value }: { name: Name; value: string }) => void;
  onToggle?: (value: boolean) => void;
  onSubmit: () => Promise<void>;
  submitCallback: () => void;
  snackMessage: string;
};

export function useFormState<Name = string>({
  setEntity,
  onToggle,
  onSubmit,
  submitCallback,
  snackMessage,
}: FormStateProps<Name>) {
  const { enqueueSnackbar } = useSnackbar();

  const handleChange = useMemo(
    () => (e: ChangeEvent<HTMLInputElement>) => {
      setEntity({
        name: e.target.name as Name,
        value: e.target.value,
      });
    },
    [setEntity],
  );

  const handleToggle = useMemo(
    () => (event: ChangeEvent<HTMLInputElement>, checked: boolean) => {
      onToggle && onToggle(checked);
    },
    [],
  );

  const handleSubmit = useMemo(
    () => async (e: FormEvent<HTMLFormElement>) => {
      e.preventDefault();
      try {
        await onSubmit();
        submitCallback();
        enqueueSnackbar(snackMessage, { variant: 'success' });
      } catch (error) {
        enqueueSnackbar('Something went wrong', { variant: 'error' });
      }
    },
    [],
  );

  return { handleChange, handleToggle, handleSubmit };
}
