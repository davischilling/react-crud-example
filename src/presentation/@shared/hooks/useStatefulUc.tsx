import { useCallback, useEffect, useLayoutEffect, useMemo, useRef, useState } from 'react';
import {
  SetState,
  StateCallback,
  StatefulUseCase,
} from '../../../domain/usecases/stateful-usecase';

type StateFulContextUseCaseProps<State, UseCaseClass extends StatefulUseCase<State>> = {
  UseCase: new (state: State, setStateCb: SetState<State>) => UseCaseClass;
  store: () => [State, (value: Partial<State>) => void];
};

function useCallbackRef<T = () => void>(callback: T) {
  const callbackRef = useRef(callback);
  useLayoutEffect(() => {
    callbackRef.current = callback;
  }, [callback]);
  return callbackRef;
}

export function useStatefulUseCase<State, UseCaseClass extends StatefulUseCase<State>>({
  UseCase,
  store,
}: StateFulContextUseCaseProps<State, UseCaseClass>) {
  const [logicState, setLogicState] = store();

  const setState = useCallback(
    (newState: any, cb?: StateCallback) => {
      setLogicState(Object.assign({}, logicState, newState));
      cb && cb();
    },
    [logicState, setLogicState],
  );

  console.log('rendering useStatefulUseCase');

  const useCaseLogicRef = useCallbackRef<UseCaseClass>(
    new UseCase(logicState, (newState, cb) => {
      setState(newState, cb);
    }),
  );

  useEffect(() => {
    return () => {
      useCaseLogicRef.current.init();
    };
  }, []);

  return { state: logicState, useCase: useCaseLogicRef.current };
}
