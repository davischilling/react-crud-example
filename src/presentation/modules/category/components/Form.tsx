import { Box, Button, Grid } from '@mui/material';
import { ChangeEvent, FormEvent } from 'react';
import { Link } from 'react-router-dom';
import { Category } from '../../../../domain/models/category';
import FormInput from '../../../@shared/components/FormInput';
import FormSwitch from '../../../@shared/components/FormSwitch';

type FormProps = {
  category: Partial<Category>;
  isDisabled: boolean;
  isLoading?: boolean;
  handleToggle: (e: ChangeEvent<HTMLInputElement>) => void;
  handleChange: (e: ChangeEvent<HTMLInputElement>) => void;
  handleSubmit: (e: FormEvent<HTMLFormElement>) => Promise<void>;
};

export default function Form({
  category,
  isDisabled,
  handleToggle,
  handleChange,
  handleSubmit,
}: FormProps) {
  return (
    <Box p={2}>
      <form onSubmit={handleSubmit}>
        <Grid container spacing={3}>
          <FormInput
            fullWidth
            required
            name="name"
            label="Name"
            value={category.name}
            onChange={handleChange}
          />
          <FormInput
            fullWidth
            required
            name="description"
            label="Description"
            value={category.description}
            onChange={handleChange}
          />

          <FormSwitch
            label="active"
            name="is_active"
            color="secondary"
            onChange={handleToggle}
            checked={category.is_active}
            inputProps={{ 'aria-label': 'controlled' }}
          />

          <Grid item xs={12}>
            <Box display="flex" gap={2}>
              <Button variant="contained" component={Link} to="/categories">
                Back
              </Button>

              <Button variant="contained" color="secondary" type="submit" disabled={isDisabled}>
                Save
              </Button>
            </Box>
          </Grid>
        </Grid>
      </form>
    </Box>
  );
}
