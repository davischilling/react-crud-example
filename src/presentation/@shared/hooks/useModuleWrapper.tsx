import React from 'react';
import { StatefulUseCase } from '../../../domain/usecases/stateful-usecase';
import createGenericContext from '../contexts/createGenericContext';
import { useStatefulUseCase } from './useStatefulUc';

type ModuleWrapperProps<State, UseCase extends StatefulUseCase<State>> = {
  DEFAULT_STATE: State;
  UseCase: new (
    state: State,
    setStateCb: (value: Partial<State>, cb?: () => void) => void,
  ) => UseCase;
};

export function useModuleWrapper<State, UseCase extends StatefulUseCase<State>>({
  DEFAULT_STATE,
  UseCase,
}: ModuleWrapperProps<State, UseCase>) {
  const { Provider, useStore } = createGenericContext<State>(DEFAULT_STATE);
  const useModuleStore = () => useStore<State>((store) => store);
  const useModuleUseCase = () =>
    useStatefulUseCase<State, UseCase>({
      UseCase,
      store: useModuleStore,
    });
  return { Provider, useModuleUseCase };
}
