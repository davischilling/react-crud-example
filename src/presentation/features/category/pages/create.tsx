import { Box, Paper, Typography } from '@mui/material';
import Form from '../components/Form';

export function CreateCategoryPage() {
  return (
    <Box>
      <Paper>
        <Box p={2}>
          <Box mb={2}>
            <Typography variant="h4">Create Category</Typography>
          </Box>
        </Box>

        <Form
          category={{}}
          isDisabled={false}
          handleToggle={() => console.log('handleToggle')}
          handleChange={() => console.log('handleChange')}
          handleSubmit={async () => console.log('handleSubmit')}
        />
      </Paper>
    </Box>
  );
}
