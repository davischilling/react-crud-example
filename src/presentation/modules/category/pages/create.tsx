import { Box, Paper, Typography } from '@mui/material';
import { useNavigate } from 'react-router-dom';
import {
  CreateCategoryUseCase,
  DEFAULT_STATE,
  State,
} from '../../../../domain/usecases/category/create';
import { useFormState } from '../../../@shared/hooks/useFormState';
import { useStatefulUseCase } from '../../../@shared/hooks/useStatefulUc';
import Form from '../components/Form';

export function CreateCategoryPage() {
  const { state, useCase } = useStatefulUseCase<State, CreateCategoryUseCase>({
    UseCase: CreateCategoryUseCase,
    DEFAULT_STATE,
  });
  const navigate = useNavigate();
  const { handleChange, handleSubmit } = useFormState({
    setEntity: useCase.setCategory,
    onSubmit: useCase.createCategory,
    submitCallback: () => navigate('/categories'),
  });

  return (
    <Box>
      <Paper>
        <Box p={2}>
          <Box mb={2}>
            <Typography variant="h4">Create Category</Typography>
          </Box>
        </Box>

        <Form
          category={state.category}
          isDisabled={false}
          handleToggle={useCase.toggleIsActive}
          handleChange={handleChange}
          handleSubmit={handleSubmit}
        />
      </Paper>
    </Box>
  );
}
