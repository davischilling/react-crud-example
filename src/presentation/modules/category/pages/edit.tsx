import { Box, Paper, Typography } from '@mui/material';
import { useNavigate, useParams } from 'react-router-dom';

import {
  DEFAULT_STATE,
  EditCategoryUseCase,
  State,
} from '../../../../domain/usecases/category/edit';
import { useFormState } from '../../../@shared/hooks/useFormState';
import { useStatefulUseCase } from '../../../@shared/hooks/useStatefulUc';
import Form from '../components/Form';

export function EditCategoryPage() {
  const id = useParams().id;

  const { state, useCase } = useStatefulUseCase<State, EditCategoryUseCase>({
    UseCase: EditCategoryUseCase,
    DEFAULT_STATE,
    INITIAL_STATE: {
      category: {
        id,
        name: '',
        description: '',
        is_active: true,
      },
    },
  });

  const navigate = useNavigate();

  const { handleChange, handleToggle, handleSubmit } = useFormState({
    setEntity: useCase.setCategory,
    onToggle: useCase.toggleIsActive,
    onSubmit: useCase.updateCategory,
    submitCallback: () => navigate('/categories'),
    snackMessage: 'Category updated successfully',
  });

  return (
    <Box>
      <Paper>
        <Box p={2}>
          <Box mb={2}>
            <Typography variant="h4">Edit Category</Typography>
          </Box>
        </Box>

        <Form
          isLoading={state.isLoading}
          category={state.category}
          isDisabled={state.isLoading}
          handleToggle={handleToggle}
          handleChange={handleChange}
          handleSubmit={handleSubmit}
        />
      </Paper>
    </Box>
  );
}
