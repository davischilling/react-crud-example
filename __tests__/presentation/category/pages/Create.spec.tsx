import { act, fireEvent, render, screen, waitFor } from '@testing-library/react';
import { BrowserRouter, useNavigate } from 'react-router-dom';
import * as useStatefulUseCaseModule from '../../../../src/presentation/@shared/hooks/useStatefulUc';
import Form from '../../../../src/presentation/modules/category/components/Form';
import { CreateCategoryPage } from '../../../../src/presentation/modules/category/pages/create';

jest.mock('../../../../src/presentation/modules/category/components/Form', () => ({
  __esModule: true,
  default: jest.fn(({ handleSubmit, handleToggle }) => (
    <div>
      Mocked Form Content
      <form onSubmit={handleSubmit}>
        <button onClick={handleToggle}>toggle</button>
        <button type="submit">submit</button>
      </form>
    </div>
  )),
}));

jest.mock('react-router-dom', () => ({
  useNavigate: jest.fn(),
}));

describe('CreateCategoryPage', () => {
  let fragment: DocumentFragment;
  let useStatefulUseCase: jest.SpyInstance;

  let navigateSpy: jest.SpyInstance;

  let createCategorySpy: jest.SpyInstance;
  let setCategorySpy: jest.SpyInstance;
  let toggleIsActiveSpy: jest.SpyInstance;

  const mockState = { category: { name: 'Mocked Category' } };

  beforeEach(async () => {
    navigateSpy = jest.fn();
    (useNavigate as jest.Mock).mockReturnValue(navigateSpy);

    createCategorySpy = jest.fn();
    setCategorySpy = jest.fn();
    toggleIsActiveSpy = jest.fn();

    useStatefulUseCase = jest
      .spyOn(useStatefulUseCaseModule as any, 'useStatefulUseCase')
      .mockReturnValue({
        state: mockState,
        useCase: {
          createCategory: createCategorySpy,
          setCategory: setCategorySpy,
          toggleIsActive: toggleIsActiveSpy,
        },
      });
    fragment = await waitFor(() => {
      const { asFragment } = render(<CreateCategoryPage />, { wrapper: BrowserRouter });
      return asFragment();
    });
  });

  test('renders the "Create Category" page', () => {
    const titleElement = screen.getByText('Create Category');
    expect(titleElement).toBeInTheDocument();
    expect(fragment).toMatchSnapshot();
  });

  test('should correctly call useNavigate', () => {
    expect(useNavigate).toHaveBeenCalledTimes(1);
    expect(useNavigate).toHaveBeenCalledWith();
  });

  test('should call navigate and onSubmit when handleSubmit returned by useFormState hook is called', async () => {
    const formSubmitButton = screen.getByText('submit');

    await act(async () => {
      fireEvent.click(formSubmitButton);
    });

    expect(createCategorySpy).toHaveBeenCalledTimes(1);
    expect(createCategorySpy).toHaveBeenCalledWith();
    expect(navigateSpy).toHaveBeenCalledTimes(1);
    expect(navigateSpy).toHaveBeenCalledWith('/categories');
  });

  test('passes the correct props to the Form component', () => {
    expect(Form).toHaveBeenCalledWith(
      expect.objectContaining({
        category: mockState.category,
        isDisabled: false,
        handleToggle: expect.any(Function),
        handleChange: expect.any(Function),
        handleSubmit: expect.any(Function),
      }),
      {},
    );
  });
});
