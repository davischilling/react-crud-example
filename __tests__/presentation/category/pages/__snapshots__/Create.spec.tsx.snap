// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CreateCategoryPage renders the "Create Category" page 1`] = `
<DocumentFragment>
  <div
    class="MuiBox-root css-0"
  >
    <div
      class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 css-1ps6pg7-MuiPaper-root"
    >
      <div
        class="MuiBox-root css-19midj6"
      >
        <div
          class="MuiBox-root css-1qm1lh"
        >
          <h4
            class="MuiTypography-root MuiTypography-h4 css-5lbw0b-MuiTypography-root"
          >
            Create Category
          </h4>
        </div>
      </div>
      <div>
        Mocked Form Content
        <form>
          <button>
            toggle
          </button>
          <button
            type="submit"
          >
            submit
          </button>
        </form>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
